lib/rucio/core/rule.py

```diff
@@ -2693,9 +3168,26 @@ def __evaluate_did_detach(eval_did: models.DataIdentifier, *, session: "Session"
         parent_dids = rucio.core.did.list_all_parent_dids(scope=eval_did.scope, name=eval_did.name, session=session)
 
         # Get all RR from parents and eval_did
-        rules = session.query(models.ReplicationRule).filter_by(scope=eval_did.scope, name=eval_did.name).with_for_update(nowait=True).all()
+        stmt = select(
+            models.ReplicationRule
+        ).where(
+            and_(models.ReplicationRule.scope == eval_did.scope,
+                 models.ReplicationRule.name == eval_did.name)
+        ).with_for_update(
+            nowait=True
+        )
+        rules = list(session.execute(stmt).scalars().all())
         for did in parent_dids:
-            rules.extend(session.query(models.ReplicationRule).filter_by(scope=did['scope'], name=did['name']).with_for_update(nowait=True).all())
+            stmt = select(
+                models.ReplicationRule
+            ).where(
+                and_(models.ReplicationRule.scope == did['scope'],
+                     models.ReplicationRule.name == did['name'])
+            ).with_for_update(
+                nowait=True
+            )
+            parent_rules = list(session.execute(stmt).scalars().all())
+            rules.extend(parent_rules)
```

Comment:
1. Doesn’t `all()` already return a `list`?
2. Why the need for `parent_rules`?

Reply:
Refactored to:
`rules.extend(session.execute(stmt).scalars().all())`